### PR TITLE
fix(gpu): LayerNorm/RMSNorm normalize over gamma dims for rank > 2 (fixes GPU transformer forward)

### PR DESCRIPTION
**Pull request by ooples**
## Summary
The DirectGpu `LayerNorm` / `RMSNorm` forward paths computed the normalized size from `input.Shape._dims[0]` (`outerSize = dims[0]; normSize = Length/dims[0]`), and `LayerNormGpu` flattened dims 2+ into the normalized size. That is only correct for a **rank-2** `[rows, D]` input.

For a transformer's **rank-3 `[B, S, D]`** activations this set `normSize = S*D` and read `gamma`/`beta` (length `D`) out of bounds, so the GPU normalized over `S*D` elements and produced garbage — measured CPU-vs-GPU `max_abs_err ≈ 70` on `[2,8,48]`.

Because `LayerNormGpu` feeds `Predict` via `TryForwardGpuOptimized`, a `Transformer<float>` trained through the AiDotNet facade scored **chance accuracy on the GPU engine**: its per-position `[B,S,D]` LayerNorm was corrupted on every forward. After this fix the untrained GPU forward is **bit-identical** to the CpuEngine reference.

## Root cause
`normSize` must be derived from the **trailing normalized dims that gamma spans** (`gamma.Length`), with every leading dim folded into `outerSize` — exactly what `CpuEngine.LayerNorm` does. The resident compiled path (`TryLayerNormResidentInto`) and the float-specialized tape path already did this; the eager generic `LayerNorm` override, `RMSNorm`, and `LayerNormGpu` still used `dims[0]`.

## Fix
- `LayerNorm<T>` override: `normSize = gamma.Length; outerSize = input.Length / normSize`.
- `RMSNorm<T>` override: same.
- `LayerNormGpu<T>` (GPU-resident Predict path): `normalizedSize = gamma.Length; batchSize = input.Length / normalizedSize` (drop the flatten-dims-2+ logic).

## Verification (RTX 3080, CUDA backend)
- Op-level CPU-vs-CUDA diff on `[2,8,48]`: `max_abs_err` **69.6 → 2.4e-7**.
- Untrained `Transformer<float>` (d48/L2/h4, SequenceClassification) forward `Predict`: **bit-identical** GPU vs CPU after the fix (garbage before).
- New regression test `LayerNorm_RankGe2_GpuMatchesCpu` (ranks 2/3/4) covering both `TensorLayerNorm` and the `LayerNormGpu` Predict path — all pass.

## Note
This fixes GPU **forward/inference** correctness for `[B,S,D]` LayerNorm. A separate, deeper issue remains where the same tiny transformer's **multi-step GPU training** (AiModelBuilder + AdamOptimizer tape loop) blows up the weights nondeterministically even though each single-batch gradient is deterministic and bit-matches the CPU engine — that is a distinct GPU-residency/weight-sync bug under the training loop's allocation churn, not addressed here.

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed GPU layer normalization for higher-rank tensors so normalization now uses the correct last-dimension size, matching CPU results and avoiding incorrect output on 3D+ inputs.

* **Tests**
  * Added regression coverage to verify GPU and CPU layer normalization produce the same results across rank-2 to rank-4 shapes.
  * Added validation for both CPU-backed and GPU-resident tensor paths.

### CODE REVIEW

File: src/AiDotNet.Tensors/Engines/DirectGpuTensorEngine.cs [hunk 1]
```diff
@@ -12580,14 +12580,15 @@ Tensor<T> IEngine.InstanceNorm<T>(Tensor<T> input, Tensor<T> gamma, Tensor<T> be
...
-        }
+        // LayerNorm normalizes over the trailing dims that gamma spans (gamma.Length), treating every
+        // leading position as an independent sample. The previous code took normalizedSize = shape[1]
+        // (times shape[2..]) with batchSize = shape[0], which is only correct for a rank-2 [B, D] input.
+        // For a transformer's rank-3 [B, S, D] activations that flattened S into the normalized size
+        // (normalizedSize = S*D) and read gamma/beta (length D) out of bounds → garbage output. This
+        // path feeds Predict (TryForwardGpuOptimized), so held-out accuracy was measured on corrupted
+        // activations even after training. Derive the sizes from gamma (matches CpuEngine.LayerNorm).
+        int normalizedSize = gamma.Length > 0 ? gamma.Length : (shape.Length > 1 ? shape[shape.Length - 1] : shape[0]);
+        int batchSize = input.Length / normalizedSize;
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/DirectGpuTensorEngine.cs:12612]: Fixed in 6e9cf76a. Added an explicit divisibility guard before the floor-division in `LayerNormGpu`: it now throws `ArgumentException` when `input.Length % normalizedSize != 0` (or `normalizedSize <= 0`), instead of silently dropping the trailing sample(s) into a mismatched launch geometry. `normalizedSize` is derived from `gamma.Length` (matching `CpuEngine.LayerNorm`).

File: tests/AiDotNet.Tensors.Tests/Engines/DirectGpu/GpuCpuCorrectnessTests.cs [hunk 2]
```diff
@@ -112,6 +112,42 @@ private void AssertGpuMatchesCpu(Tensor<float> gpu, Tensor<float> cpu, string op
...
+        var cpu = _cpu.TensorLayerNorm(input, gamma, beta, 1e-5);
+        var gpu = _gpu.TensorLayerNorm(input, gamma, beta, 1e-5);
+        AssertGpuMatchesCpu(gpu, cpu, $"TensorLayerNorm[{string.Join(",", shape)}]");
+
+        // The GPU-resident forward path (LayerNormGpu, used by Predict/TryForwardGpuOptimized)
+        // must normalize identically for rank >= 3. It requires a GPU-resident input.
+        var gpuInput = input.Gpu();
+        var (gpuResident, _, _) = _gpu.LayerNormGpu(gpuInput, gamma, beta, 1e-5);
+        AssertGpuMatchesCpu(gpuResident, cpu, $"LayerNormGpu[{string.Join(",", shape)}]");
+    }
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tensors.Tests/Engines/DirectGpu/GpuCpuCorrectnessTests.cs:170]: Done in 6e9cf76a. `LayerNorm_RankGe2_GpuMatchesCpu` now captures the previously-discarded `SaveMean`/`SaveInvVar` from the `LayerNormGpu` tuple and asserts them against a CPU recomputation (population mean and `1/sqrt(var + eps)` per gamma-spanning sample), so the backward-path statistics are covered too — not just the normalized output.
